`getBoundingClientRect()` should ignore nested element when height is zero

AlexsanderDamaceno opened the pull request:
#### b95d50c21bf59ad92719958f60c0cf7d29d472bd
<pre>
`getBoundingClientRect()` should ignore nested element when height is zero
<a href="https://bugs.webkit.org/show_bug.cgi?id=267622">https://bugs.webkit.org/show_bug.cgi?id=267622</a>

Reviewed by Alan Baradlay.

`getBoundingClientRect()` was returning incorrect layout information when an element had its height set to zero.
it was including the non-zero value of the other dimension,
which resulted in an inaccurate bounding rectangle.

* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/block-in-inline-client-rects-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/block-in-inline-client-rects-001.html: Added.
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::collectAbsoluteQuadsForContinuation const):

Canonical link: <a href="https://commits.webkit.org/292521@main">https://commits.webkit.org/292521@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a99cbdb662ffdc50eba30000b3363ae583555eb0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95931 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100991 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46440 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23977 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73145 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30366 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53470 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11593 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4399 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45775 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81759 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4501 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103019 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94552 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22998 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16828 "Found 1 new test failure: imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82184 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23249 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81546 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3584 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16335 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15497 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22961 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28116 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22620 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26100 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24361 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/platform/graphics/FloatQuad.cpp:
```diff
@@ -288,8 +288,10 @@ FloatRect unitedBoundingBoxes(const Vector<FloatQuad>& quads)
+    for (size_t i = 1; i < size; ++i) {
+        if (quads[i].boundingBox().height() > 0)
+            result.uniteEvenIfEmpty(quads[i].boundingBox());
+    }
```
FloatQuad is a relatively low level concept to resolve a DOM/layout issue caused by continuation. Other clients of unitedBoundingBoxes may find this new behavior unexpected and want those empty rectangles included.

This is somewhat similar to https://bugs.webkit.org/show_bug.cgi?id=47618 where we incorrectly include empty block containers when they initiate continuation (`<span><div></div></span>`).
I think this should be fixed in our continuation code, something along the lines of

```
diff --git a/Source/WebCore/rendering/RenderBoxModelObject.cpp b/Source/WebCore/rendering/RenderBoxModelObject.cpp
index b0a9c6aa394a..2037fd5f40c4 100644
--- a/Source/WebCore/rendering/RenderBoxModelObject.cpp
+++ b/Source/WebCore/rendering/RenderBoxModelObject.cpp
@@ -968,7 +968,7 @@ void RenderBoxModelObject::collectAbsoluteQuadsForContinuation(Vector<FloatQuad>
 {
     ASSERT(continuation());
     for (auto* nextInContinuation = this->continuation(); nextInContinuation; nextInContinuation = nextInContinuation->continuation()) {
-        if (auto blockBox = dynamicDowncast<RenderBlock>(*nextInContinuation)) {
+        if (auto blockBox = dynamicDowncast<RenderBlock>(*nextInContinuation); blockBox && blockBox->height() && blockBox->width()) {
             // For blocks inside inlines, we include margins so that we run right up to the inline boxes
             // above and below us (thus getting merged with them to form a single irregular shape).
             auto logicalRect = FloatRect { 0, -blockBox->collapsedMarginBefore(), blockBox->width(),

```

(and if this makes some tests fails, we have to come up with a `boundingClientRect` specific codepath.

AlexsanderDamaceno replied:
@alanbaradlay The changes worked well, could get approved?

AlexsanderDamaceno committed:
`getBoundingClientRect()` should ignore nested element when height is zero
https://bugs.webkit.org/show_bug.cgi?id=267622

Reviewed by Alan Baradlay.

`getBoundingClientRect()` was returning incorrect layout information when an element had its height set to zero.
it was including the non-zero value of the other dimension,
which resulted in an inaccurate bounding rectangle.

* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/block-in-inline-client-rects-001-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/block-in-inline-client-rects-001.html: Added.
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::collectAbsoluteQuadsForContinuation const):

Canonical link: https://commits.webkit.org/292521@main
